### PR TITLE
Grw 650/checkout payment

### PR DESCRIPTION
**Pull request by dendoesit**
## What?

Payment handling with Adyen for both simple cards + 3ds cards.
Implementing a new Adyen middleware to handle the new routes. 
We still need to do some better error handling + loading states. 

## Why?
New Adyen mutations require a new way of handling Adyen. Now we have a button that controls the submit of Adyen from outside the Adyen code.

CPAS

- [x] new adyen component
- [x] connect payment endpoint + test
- [x] new endpoint for payment methods instead of all available payment + test
- [x] onAdditionalDetails endpoint + test
- [x] if payment success start checkout and redirect to final page
- [x] normal card submit 
- [x] 3ds card submit
- [x] test flow end to end

## Testing
Create a new offer . In the offer page replace /offer with /checkout/details and follow the flow there. 
Test cards can be found here : https://docs.adyen.com/development-resources/testing/test-card-numbers#test-3d-secure-2-authentication

## Known Issues
Inside Checkout payment if you put the ssn before the card details the page rerenders and the adyen implementation breaks so you need to refresh to finish the flow.

### CODE REVIEW

File: src/client/pages/Checkout/CheckoutPayment/CheckoutPayment.tsx [hunk 1]
```diff
@@ -43,7 +45,7 @@ const AdyenContainer = styled.div`
       display: none;
     }
     .adyen-checkout__payment-method {
-      background: transparent;
+      background: trans ent;
```
Comment (robinandeer) [inline src/client/pages/Checkout/CheckoutPayment/CheckoutPayment.tsx]: this must be a mistake 🙃

File: src/client/pages/Checkout/CheckoutPayment/CheckoutPayment.tsx [hunk 2]
```diff
@@ -170,11 +172,21 @@ export const CheckoutPayment = ({
...
   const [startCheckout] = useStartCheckoutMutation()
+  const [addPaymentTokenMutation] = useAddPaymentTokenMutation()
   const [getStatus] = useCheckoutStatusLazyQuery({
     pollInterval: 1000,
   })
 
+  const { payment: is3DsComplete } = useParams<{ payment: string }>()
   const onConnectPaymentSuccess = useCallback(async () => {
+    const paymentTokenId = storage.session.getSession()?.paymentTokenId
+    if (!paymentTokenId) throw new Error('No token payment id')
```
Comment (robinandeer) [inline src/client/pages/Checkout/CheckoutPayment/CheckoutPayment.tsx:184]: Why not pass this as a parameter to the function instead?

File: src/server/adyenMiddleware.ts [hunk 3]
```diff
@@ -107,35 +107,33 @@ export const handleNewAdyen3dsPostRedirect: Router.IMiddleware<
...
-      ['Authorised', 'Pending'].includes(
+      ['AUTHORISED', 'PENDING'].includes(
         result.data?.data?.paymentConnection_submitAdyenRedirection?.status,
       )
     ) {
       ctx.redirect(
         ctx.params.locale
-          ? `/${ctx.params.locale}/new-member/download`
-          : '/new-member/connect-download',
+          ? `/${ctx.params.locale}/new-member/checkout/payment/${quoteCartId}/true`
```
Comment (robinandeer) [inline src/server/adyenMiddleware.ts:123]: Hmm unsure if this is the best URL to use -- what does "true" mean?

Comment (dendoesit) [reply src/server/adyenMiddleware.ts:123]: yeah, was planning to change this at a later date but lacked a better name. guess ' 3dssuccess ' would be better.

File: src/client/pages/Checkout/CheckoutPayment/CheckoutPayment.tsx (hunk 2 above)
Comment (dendoesit) [reply src/client/pages/Checkout/CheckoutPayment/CheckoutPayment.tsx]: added this as a parameter to onConnectPaymentSuccess.

File: src/client/pages/OfferNew/utils.ts [hunk 4]
```diff
@@ -337,6 +337,7 @@ export const typeOfResidenceTextKeys: Record<
   [TypeOfContract.SeApartmentStudentBrf]:
     'CHECKOUT_DETAILS_RESIDENCE_TYPE_OWN_APARTMENT',
   [TypeOfContract.SeHouse]: 'CHECKOUT_DETAILS_RESIDENCE_TYPE_OWN_HOUSE',
+  [TypeOfContract.NoHouse]: 'CHECKOUT_DETAILS_RESIDENCE_TYPE_OWN_HOUSE',
```
Comment (mehdielket) [inline src/client/pages/OfferNew/utils.ts:340]: Why did we add NoHouse to start with?

File: src/server/serverEntry.tsx [hunk 5]
```diff
@@ -122,6 +125,11 @@ router.post(
   handleAdyen3dsPostRedirect,
 )
 
+router.post(
+  localePathPattern + '/new-member/connect-payment/adyen-newcallback',
```
Comment (mehdielket) [inline src/server/serverEntry.tsx:129]: `adyen-cps-callback` or `adyen-quotecart-callback` ?

Comment (mehdielket) [reply src/server/serverEntry.tsx:129]: the `new` always ages very badly IMHO

File: src/client/pages/OfferNew/utils.ts (hunk 4 above)
Comment (dendoesit) [reply src/client/pages/OfferNew/utils.ts:340]: in order to pass typcheck. it's the same as the adtraction discussion we had on slack yesterday.

**Review by amandaflod — APPROVED** [review]
Thank you for going through everything with me! 🙌 Looks good!

**Review by robinandeer — APPROVED** [review]
Massive job @dendoesit 👏 

There's so much here it's impossible to keep track of. We really should break this up and expose as little as possible to React. But that's for a different day. Best to merge and iterate I think

**Pull request merged**